File: src/main/java/net/atos/client/zgw/zrc/model/RolMedewerker.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.client.zgw.zrc.model;

import static org.apache.commons.lang3.StringUtils.isNotBlank;

import java.net.URI;
import java.util.Objects;
import java.util.UUID;

import jakarta.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;

import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum;
import nl.info.client.zgw.zrc.model.generated.MedewerkerIdentificatie;
import nl.info.client.zgw.ztc.model.generated.RolType;

/**
 * Manually copied from {@link nl.info.client.zgw.zrc.model.generated.RolMedewerker} and modified to allow for
 * polymorphism using a generic base {@link Rol} class.
 * Ideally we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 */
public class RolMedewerker extends Rol<MedewerkerIdentificatie> {

    public RolMedewerker() {
    }

    /**
     * For testing purposes only where we need a UUID.
     */
    public RolMedewerker(
            final UUID uuid,
            final RolType roltype,
            final String roltoelichting,
            final MedewerkerIdentificatie betrokkeneIdentificatie
    ) {
        super(uuid, roltype, BetrokkeneTypeEnum.MEDEWERKER, betrokkeneIdentificatie, roltoelichting);
    }

    public RolMedewerker(
            final URI zaak,
            final RolType roltype,
            final String roltoelichting,
            // it is possible in the ZGW API to have a RolMedewerker without a Medewerker,
            // and this does occur in practice in certain circumstances
            @Nullable final MedewerkerIdentificatie betrokkeneIdentificatie
    ) {
        super(zaak, roltype, BetrokkeneTypeEnum.MEDEWERKER, betrokkeneIdentificatie, roltoelichting);
    }

    public String getNaam() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        final MedewerkerIdentificatie medewerker = getBetrokkeneIdentificatie();
        if (isNotBlank(medewerker.getAchternaam())) {
            final StringBuilder naam = new StringBuilder();
            if (isNotBlank(medewerker.getVoorletters())) {
                naam.append(medewerker.getVoorletters());
                naam.append(StringUtils.SPACE);
            }
            if (isNotBlank(medewerker.getVoorvoegselAchternaam())) {
                naam.append(medewerker.getVoorvoegselAchternaam());
                naam.append(StringUtils.SPACE);
            }
            naam.append(medewerker.getAchternaam());
            return naam.toString();
        } else {
            return medewerker.getIdentificatie();
        }
    }

    @Override
    protected boolean equalBetrokkeneIdentificatie(final MedewerkerIdentificatie identificatie) {
        final MedewerkerIdentificatie betrokkeneIdentificatie = getBetrokkeneIdentificatie();
        if (betrokkeneIdentificatie == identificatie) {
            return true;
        }
        if (identificatie == null) {
            return false;
        }
        return Objects.equals(betrokkeneIdentificatie.getIdentificatie(), identificatie.getIdentificatie());
    }

    @Override
    public String getIdentificatienummer() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        return getBetrokkeneIdentificatie().getIdentificatie();
    }

    @Override
    protected int hashCodeBetrokkeneIdentificatie() {
        if (getBetrokkeneIdentificatie() == null) {
            return -1;
        }
        return Objects.hash(getBetrokkeneIdentificatie().getIdentificatie());
    }
}

File: src/main/kotlin/nl/info/client/zgw/zrc/model/RolMedewerker.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.client.zgw.zrc.model

import jakarta.annotation.Nullable
import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum
import nl.info.client.zgw.zrc.model.generated.MedewerkerIdentificatie
import nl.info.client.zgw.ztc.model.generated.RolType
import org.apache.commons.lang3.StringUtils
import java.net.URI
import java.util.Objects
import java.util.UUID

/**
 * Manually copied from [nl.info.client.zgw.zrc.model.generated.RolMedewerker] and modified to allow for
 * polymorphism using a generic base [Rol] class.
 * Ideally, we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 */
class RolMedewerker : Rol<MedewerkerIdentificatie> {
    constructor() : super()

    /**
     * For testing purposes only where we need a UUID.
     */
    constructor(
        uuid: UUID,
        roltype: RolType,
        roltoelichting: String,
        betrokkeneIdentificatie: MedewerkerIdentificatie?
    ) : super(uuid, roltype, BetrokkeneTypeEnum.MEDEWERKER, betrokkeneIdentificatie, roltoelichting)

    constructor(
        zaak: URI,
        roltype: RolType,
        roltoelichting: String,
        // it is possible in the ZGW API to have a RolMedewerker without a Medewerker,
        // and this does occur in practice in certain circumstances
        @Nullable medewerkerIdentificatie: MedewerkerIdentificatie?
    ) : super(zaak, roltype, BetrokkeneTypeEnum.MEDEWERKER, medewerkerIdentificatie, roltoelichting)

    override val naam: String?
        get() {
            val medewerker = betrokkeneIdentificatie ?: return null
            return if (StringUtils.isNotBlank(medewerker.achternaam)) {
                buildString {
                    if (StringUtils.isNotBlank(medewerker.voorletters)) {
                        append(medewerker.voorletters).append(StringUtils.SPACE)
                    }
                    if (StringUtils.isNotBlank(medewerker.voorvoegselAchternaam)) {
                        append(medewerker.voorvoegselAchternaam).append(StringUtils.SPACE)
                    }
                    append(medewerker.achternaam)
                }
            } else {
                medewerker.identificatie
            }
        }

    @Suppress("ReturnCount")
    override fun equalBetrokkeneIdentificatie(identificatie: MedewerkerIdentificatie?): Boolean {
        val ownIdentificatie = betrokkeneIdentificatie
        if (ownIdentificatie === identificatie) {
            return true
        }
        if (identificatie == null) {
            return false
        }
        return Objects.equals(ownIdentificatie?.identificatie, identificatie.identificatie)
    }

    override val identificatienummer: String?
        get() = betrokkeneIdentificatie?.identificatie

    override fun hashCodeBetrokkeneIdentificatie(): Int {
        val identificatie = betrokkeneIdentificatie ?: return -1
        return Objects.hash(identificatie.identificatie)
    }
}
